src/Build/Globbing/MSBuildGlob.cs

```diff
@@ -0,0 +1,271 @@
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+//-----------------------------------------------------------------------
+
+
+using System;
+using System.IO;
+using System.Text.RegularExpressions;
+using Microsoft.Build.Shared;
+
+namespace Microsoft.Build.Globbing
+{
+    /// <summary>
+    ///     Represents a parsed MSBuild glob.
+    ///     An MSBuild glob is composed of three parts:
+    ///     - fixed directory part: "a/b/" in "a/b/**/*test*/**/*.cs"
+    ///     - wildcard directory part: "**/*test*/**/" in "a/b/**/*test*/**/*.cs"
+    ///     - file name part: "*.cs" in "a/b/**/*test*/**/*.cs"
+    /// </summary>
+    public class MsBuildGlob : IMsBuildGlob
+    {
+        //private readonly Lazy<string[]> _expandedFiles;
+        internal readonly string _fileSpec;
+        internal readonly string _globRoot;
+        private readonly string _matchFileExpression;
+        internal readonly bool _needsRecursion;
+        private readonly Lazy<Regex> _regex;
+
+        /// <summary>
+        ///     The fixed directory part.
+        /// </summary>
+        public string FixedDirectoryPart { get; }
+
+        /// <summary>
+        ///     The wildcard directory part
+        /// </summary>
+        public string WildcardDirectoryPart { get; }
+
+        /// <summary>
+        ///     The file name part
+        /// </summary>
+        public string FilenamePart { get; }
+
+        /// <summary>
+        ///     Whether the glob was parsed sucsesfully from a string.
+        ///     Illegal glob strings contain:
+        ///     - invalid path characters (other than the wildcard characters themselves)
+        ///     - "..."
+        ///     - ":"
+        ///     In addition, the wildcard directory part:
+        ///     - cannot contain ".."
+        ///     - if ** is present it must appear alone between slashes
+        /// </summary>
+        public bool IsLegal { get; }
+
+        private MsBuildGlob(
+            string globRoot,
+            string fileSpec,
+            string fixedDirectoryPart,
+            string wildcardDirectoryPart,
+            string filenamePart,
+            string matchFileExpression,
+            bool needsRecursion,
+            bool isLegalFileSpec)
+        {
+            FixedDirectoryPart = fixedDirectoryPart;
+            WildcardDirectoryPart = wildcardDirectoryPart;
+            FilenamePart = filenamePart;
+            IsLegal = isLegalFileSpec;
+
+            _globRoot = globRoot;
+            _fileSpec = fileSpec;
+            _matchFileExpression = matchFileExpression;
+            _needsRecursion = needsRecursion;
+
+            // compile the regex since it's expected to be used multiple times
+            _regex =
+                new Lazy<Regex>(
+                    () => new Regex(_matchFileExpression, FileMatcher.DefaultRegexOptions | RegexOptions.Compiled),
+                    true);
```

Comment:
Docs suggest true enables thread safety: https://msdn.microsoft.com/en-us/library/dd783092(v=vs.110).aspx
They are a bit confusing, so this is what the code suggests (if I followed all the layers of indirection correctly)

True leads to this construction code which uses a lock: https://source.dot.net/#System.Private.CoreLib/shared/System/Lazy.cs,f8f9288e02c57df5

False leads to this one which does not use a lock:
https://source.dot.net/#System.Private.CoreLib/shared/System/Lazy.cs,f3a8e7f97a578b26,references

So I guess setting it to true enables thread safety.